### ISSUE TITLE
Magic DNS upstream load balancing not working on iOS

### ISSUE THREAD

**Issue by FiloSottile** [tailscale/tailscale#2663]
I had two Magic DNS upstreams set up (with override system DNS on), both Tailscale nodes. The first happened to be down. My macOS machine was working fine, presumably picking the other one as "faster" than the dead one. My iOS device instead was unable to resolve names, and getting stuck loading pages. As soon as I removed the broken upstream, the iOS device started to load pages correctly.

**Comment by danderson** [comment]
Do you also have MagicDNS enabled, or only upstream resolvers set? Can you share the Tailscale IP for the iOS device, so I can peek at debug logs? (either here or to support@tailscale.com)

**Comment by FiloSottile** [comment]
I have MagicDNS on. The device is 100.97.8.94.

**Comment by bradfitz** [comment]
Weird. macOS and iOS share basically all the same code paths. I wouldn't expect them to differ.

**Comment by danderson** [comment]
Yeah, same thought. With MagicDNS+upstreams enabled, we proxy all DNS traffic through quad-100, and from there the forwarder broadcasts to all upstreams. In the forwarder code, the only ios-specific tweak is to limit the in-flight request count when talking to well-known DoH servers (quad-1, quad-8, quad-9), to save memory.

We're going to need more debug spam to diagnose. Unfortunately, that's the kind of thing that tends to make iOS OOMkill us :/

Next steps would be someone with iOS to reproduce this, so we can attach a debugger and figure out what's happening. The setup in question here has 2 upstream resolvers, both Tailscale IPs.

**Comment by mayakacz** [comment]
@FiloSottile Are you able to reproduce this? Would like to get more information to debug.

**Comment by DentonGentry** [comment]
In Tailscale 1.38 we're expecting to remove the limit on max in-flight DoH queries altogether, and rely on backpressuring the platform while it waits for earlier DNS lookups to complete. The limit was originally put in place for the benefit of iOS13 and iOS14 with their limited memory available for Network Extensions, but is no longer needed.

**Comment by DentonGentry** [comment]
I think this was likely a symptom of the small number of in-flight DNS requests on iOS Tailscale releases prior to 1.38.

Barring further comment, we'll expect to close this during the next regular bug scrub.
